### PR TITLE
Binders with default settings inherit from module's configuration

### LINKED ISSUE #229: RabbitMessageChannelBinderConfiguration no longer using externally defined ConnectionFactory
In the Autowire of the `ConnectionFactory rabbitConnectionFactory` an externally defined ConnectionFactory is no longer used.  Only the internally created ConnectionFactory is considered.

### PR DESCRIPTION
**Pull request by mbogoevici**
Resolves #229
Optimizes connection factory creation for binders that can use Boot's autoconfiguration

### CODE REVIEW

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java [hunk 1]
```diff
@@ -49,8 +51,8 @@ public DefaultBinderFactory(Map<String, BinderConfiguration> binderConfiguration
 	}
 
 	@Override
-	public void setEnvironment(Environment environment) {
-		this.environment = environment;
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: minor thing, but this doesn't need the `throws` clause

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java [hunk 2]
```diff
@@ -110,12 +113,14 @@ else if (binderConfigurations.size() == 1) {
 				defaultDomain += ".";
 			}
 			args.add("--spring.jmx.default-domain=" + defaultDomain + "binder." + configurationName);
-			args.add("--spring.application.admin.enabled=false");
 			SpringApplicationBuilder springApplicationBuilder =
 					new SpringApplicationBuilder()
 							.sources(binderConfiguration.getBinderType().getConfigurationClasses())
 							.bannerMode(Mode.OFF)
 							.web(false);
+			if (binderProperties.isEmpty() && context != null) {
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: Might be worth a brief comment here to explain why we only set the parent context in this case.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java (hunk 1 above)
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: I blame IntelliJ :)

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java (hunk 2 above)
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: Will add.

**Comment by mbogoevici** [comment]
Rebased, added changes in https://github.com/spring-cloud/spring-cloud-stream/issues/232 and, additionally added support for suppressing environment inheritance (wherever necessary).

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java [hunk 3]
```diff
@@ -110,19 +113,30 @@ else if (binderConfigurations.size() == 1) {
...
+			// If the environment is not customized and a main context is available, we will set the latter as parent.
+			// This ensures that the defaults and user-defined customizations (e.g. custom connection factory beans)
+			// are propagated to the binder context. If the environment is customized, then the binder context should
+			// not inherit any beans from the parent
+			if (binderProperties.isEmpty() && context != null) {
+				springApplicationBuilder.parent(context);
+			}
+			else if (environment != null && binderConfiguration.isInheritEnvironment()) {
+				StandardEnvironment binderEnvironment = new StandardEnvironment();
+				environment.merge(environment);
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: This looks like a typo. I assume the intent is to merge into the `binderEnvironment`?

Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: Yeah.

File: spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binder/BinderFactoryConfigurationTests.java [hunk 4]
```diff
@@ -90,6 +89,29 @@ public void loadBinderTypeRegistryWithOneBinder() throws Exception {
 	}
 
 	@Test
+	public void loadBinderTypeRegistryWithOneBinderAndSharedEnvironment() throws Exception {
```
Comment (markfisher) [inline spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binder/BinderFactoryConfigurationTests.java]: this test and `loadBinderTypeRegistryWithOneBinderAndIsolatedEnvironment` make the same assertions - shouldn't we be testing something that varies with the boolean property?

Comment (mbogoevici) [inline spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binder/BinderFactoryConfigurationTests.java]: Actually, they're not supposed to. Updating the tests now.

File: spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.data.redis.connection.RedisConnectionFactory;
+
+/**
+ * @author Marius Bogoevici
+ */
+public class RedisBinderModuleTests {
+
+	@ClassRule
+	public static RedisTestSupport rabbitTestSupport = new RedisTestSupport();
```
Comment (markfisher) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java]: wrong param name (copy paste) ;)

File: spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	public void testParentConnectionFactoryInheritedByDefault() {
+		context = SpringApplication.run(SimpleProcessor.class);
+		BinderFactory binderFactory = context.getBean(BinderFactory.class);
+		Binder binder = binderFactory.getBinder(null);
+		assertThat(binder, instanceOf(RedisMessageChannelBinder.class));
+		DirectFieldAccessor binderFieldAccessor = new DirectFieldAccessor(binder);
+		RedisConnectionFactory binderConnectionFactory =
+				(RedisConnectionFactory) binderFieldAccessor.getPropertyValue("connectionFactory");
+		assertThat(binderConnectionFactory, instanceOf(RedisConnectionFactory.class));
+		RedisConnectionFactory connectionFactory = context.getBean(RedisConnectionFactory.class);
```
Comment (markfisher) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java]: in the rabbit tests, the class passed to `getBean` is `ConnectionFactory.class` - why is this different?

File: spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	public void testParentConnectionFactoryInheritedIfOverridden() {
+		context = new SpringApplication(SimpleProcessor.class, ConnectionFactoryConfiguration.class).run();
+		BinderFactory binderFactory = context.getBean(BinderFactory.class);
+		Binder binder = binderFactory.getBinder(null);
+		assertThat(binder, instanceOf(RedisMessageChannelBinder.class));
+		DirectFieldAccessor binderFieldAccessor = new DirectFieldAccessor(binder);
+		RedisConnectionFactory binderConnectionFactory =
+				(RedisConnectionFactory) binderFieldAccessor.getPropertyValue("connectionFactory");
+		assertThat(binderConnectionFactory, is(MOCK_CONNECTION_FACTORY));
+		RedisConnectionFactory connectionFactory = context.getBean(RedisConnectionFactory.class);
```
Comment (markfisher) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java]: same question as in comment on line 76

File: spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		params.add("--spring.cloud.stream.binders.custom.type=redis");
+		params.add("--spring.cloud.stream.binders.custom.environment.foo=bar");
+		context = SpringApplication.run(SimpleProcessor.class, params.toArray(new String[]{}));
+		BinderFactory binderFactory = context.getBean(BinderFactory.class);
+		Binder binder = binderFactory.getBinder(null);
+		assertThat(binder, instanceOf(RedisMessageChannelBinder.class));
+		DirectFieldAccessor binderFieldAccessor = new DirectFieldAccessor(binder);
+		RedisConnectionFactory binderConnectionFactory =
+				(RedisConnectionFactory) binderFieldAccessor.getPropertyValue("connectionFactory");
+		RedisConnectionFactory connectionFactory = context.getBean(RedisConnectionFactory.class);
```
Comment (markfisher) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java]: ditto

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java [hunk 9]
```diff
@@ -51,8 +51,8 @@ public DefaultBinderFactory(Map<String, BinderConfiguration> binderConfiguration
 	}
 
 	@Override
-	public void setEnvironment(Environment environment) {
-		this.environment = environment;
+	public void setApplicationContext(ApplicationContext applicationContext) {
```
Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java]: should this have an `Assert.instanceOf(..)` prior to the cast?

**Comment by markfisher** [comment]
will address my 2 last comments while merging

**Comment by markfisher** [comment]
merged along with https://github.com/spring-cloud/spring-cloud-stream/commit/df1817cc518eb979d49ab1030fada5ec4091cd8a

**Pull request merged**